types/Promises.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+use React\Promise\FulfilledPromise;
+use React\Promise\PromiseInterface;
+use Throwable;
+
+use function PHPStan\Testing\assertType;
+use function React\Promise\all;
+use function React\Promise\any;
+use function React\Promise\race;
+use function React\Promise\reject;
+use function React\Promise\resolve;
+
+$passThroughBoolFn = static fn (bool $bool): bool => $bool;
+$passThroughThrowable = static function (Throwable $t): PromiseInterface {
+    return reject($t);
+};
+$stringOrInt = function (): int|string {
+  return time() % 2 ? 'string' : time();
+};
+$tosseable = new Exception('Oops I did it again!');
+
+/**
+ * basic
+ */
+assertType('React\Promise\PromiseInterface<bool>', resolve(true));
+assertType('React\Promise\PromiseInterface<int|string>', resolve($stringOrInt()));
+assertType('React\Promise\PromiseInterface<bool>', resolve(resolve(true)));
+
+/**
+ * chaining
+ */
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then()->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then(null)->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then($passThroughBoolFn)->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then($passThroughBoolFn, $passThroughThrowable)->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then(null, $passThroughThrowable)->then($passThroughBoolFn));
+assertType('React\Promise\PromiseInterface<bool>', resolve(true)->then()->then(null, $passThroughThrowable)->then($passThroughBoolFn));
+
+/**
+ * all
+ */
+assertType('React\Promise\PromiseInterface<array<bool>>', all([resolve(true), resolve(false)]));
+assertType('React\Promise\PromiseInterface<array<bool>>', all([resolve(true), false]));
+assertType('React\Promise\PromiseInterface<array<bool|int>>', all([true, time()]));
+assertType('React\Promise\PromiseInterface<array<bool|int>>', all([resolve(true), resolve(time())]));
+assertType('React\Promise\PromiseInterface<array<bool|float>>', all([resolve(true), hrtime()]));
+assertType('React\Promise\PromiseInterface<array<bool|int>>', all([true, resolve(time())]));
```

Comment:
Perhaps it would be fine to comment these out for now. Most of the time people will be using homogeneous lists anyway. And if someone needs to use a heterogeneous one, they can always add it to ignore list in `phpstan.neon`.

Reply:
Lines 46 and 47 work fine, lines 48 and 49 don't. If the bug in PHPStan doesn't get fixed before this PR will be merged we'll comment 48 and 49 out. As that's not a very common situation. 46 and maybe 47 I have a lot in applications where I don't directly control the typing and accept multiple. An example is the GitHub client I've been generating having a [`Repository`](https://github.com/php-api-clients/github/blob/v0.2.x/src/Schema/Repository.php) and a [`SimpleRepository`](https://github.com/php-api-clients/github/blob/v0.2.x/src/Schema/SimpleRepository.php). Both work for where I use them and it gets either depending on where something originates from. So something returning an array with both is perfectly possible.